### ISSUE TITLE
Update EX Extension Granted & Create COS Payment Automation (Passed PROD UAT)

### ISSUE THREAD

**Issue by TracyLinder** [cityofaustin/atd-data-tech#7631]
### Updates to Right-of-Way Admin Process - Extension Granted Attempt

**Overview:**
In order to implement Cost of Service payment automation on excavation extension fees, the workflow and features of the AMANDA extension features must be updated. 

**Workflow / Acceptance Criteria:**

Role | Task
-- | --
User | Saves attempt 'Extension Granted'
System | Performs Validations:✔️Process Infos are entered: New Start Date, New End Date✔️Folder issue date must exist✔️ROW Contractor must exist & be active
System | Updates Right-of-Way Admin Process infos values to NULL for next extension.
System | Inserts extenion comment in folder description.
System | Copies process infos to folder infos: - New Start Date  ➡️Extension Start Date - New End Date ➡️ Extension End Date
System: | ❓Decision: Is this Franchise, COS, or Private/Project Connect?
  |  
  | 🔹 FRANCHISE 🔹
System | ✔️Update folder status to Active✔️Update folder expiration date to info 'Extension End Date'
  |  
  | 🔹 COS 🔹
System | ❓Decision: Is info Secondary yes or no?
System | IF YES:   ✔️Insert Fees: Excavation Permit Fee - Secondary, ATD Tech Surcharge ROW✔️Create bill✔️Pay bill with COS payment type - (A.U.P. triggered by payment)
System | IF NO:   ✔️Insert Fees: Excavation Permit Fee - Extension, ATD Tech Surcharge ROW✔️Create bill✔️Pay bill with COS payment type - (A.U.P. triggered by payment)
  |  
  | 🔹 PRIVATE DEVELOPER / PROJECT CONNECT 🔹
System | ❓Decision: Is info Secondary yes or no?
System | IF YES:✔️Insert Fees: Excavation Permit Fee - Secondary, ROW Usage Fee(s), ATD Tech Surcharge ROW - (do not create bill)
System | IF NO:✔️Insert Fees: Excavation Permit Fee - Extension, ROW Usage Fee(s), ATD Tech Surcharge ROW - (do not create bill)
  | ⛔️ STOP/DONE

----

![Excavation Extension Granted](https://user-images.githubusercontent.com/49930868/146278522-93f6fd94-876f-41b0-a10b-3a64380cad8a.jpg)

![Activate Upon Payment](https://user-images.githubusercontent.com/49930868/146278559-8f285466-2b6b-4bcd-b257-5287aa9de448.jpg)

**Comment by TracyLinder** [comment]
@vschandramouli is going to meet w/ stakeholders and finalize the AC

**Comment by vschandramouli** [comment]
Updated AC, moving to Backlog. Placing previous notes from the ticket in this comment:

**Current Process:**

1. Staff manually inserts & bills the EX extension fees. 
2. Customer pays the bill & notifies ROW staff by email
3. ROW Staff saves 'Extension Granted' on ROW Admin Process
4. System updates folder status to Active, updates expiration date, and probably a couple of other things

**Future Process:**
1. Staff saves Extension Granted' on ROW Admin Process
2. System inserts & bills the correct extension fees
    - Private Development & Project Connect may have to pay ROW Usage fees, while the others (COS Departments) do not
3. The bill is paid, either by the customer or by a COS automated payment.
4. System updates folder status to Active, updates expiration date, and probably a couple of other things

**Comment by TracyLinder** [comment]
@Nadin-Nader Can this go on deck next week?

**Comment by TracyLinder** [comment]
AC has been updated & confirmed by Kim & Paloma.

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan I added the new AC to the top per our 12/10 meeting with the stakeholders:

Add 2 New Infos (Extension start (75993) and Extension end (75994) dates)
The New Infos (Extension start (75993) and Extension end (75994) dates) should updated by the process info
The two new info fields:
(Extension Start Date (75993)
(Extension End Date (75994)

**Comment by TracyLinder** [comment]
Task for me: Update AC to include #7284 & #7909

**Comment by TracyLinder** [comment]
## These are the issue descriptions prior to the update I'm doing:


### This Issue #7631:

**Update the EX Extension workflow to mirror the TURP extension workflow.** 

**Acceptance Criteria**
- The AC for this item has been split into different functionalities as there are multiple items that need to be addressed as part of this effort. 

**Fee Insertion for EX Extensions**
System should insert fees and bill the fees in the EX folders when 'Extension Granted' attempt is made in the ROW Admin Process based on the following criteria: 

**Secondary/Extension Fees**
-- If Info field 'Secondary?' is answered Yes, insert Excavation Permit Fee - Secondary ($95) 
-- If Info field 'Secondary?' is answered No, insert Excavation Permit Fee - Extension ($195)

**ROW Usage Fee** 
- System should use the existing logic to determine if the ROW Usage Fee should be inserted. Fee should insert when Extension Granted is saved. 
_Existing Logic:_
If any of the infos below fields contain a non-zero value (any value apart from blank or 0), system should calculate and insert ROW Usage fee using the existing logic. 

- 76622 Lanes Affected Days
- 76621 Number of Lanes Affected
- 76624 Structures Affected Days
- 76623 # of Concrete Structures on ROW

-  _ATD doesn't charge ROW Usage fees to COS Departments; ROW chose not to program around this, instead they will leave the values at 0 so no ROW Usage fee is added._

**ATD Tech Surcharge**
-- System should also insert ATD tech surcharge last automatically, which 4% of any fees inserted from the list above. (Taken care of on #7707)

**Activate Upon Payment Functionality after Extension**
_Some of this functionality may have been addressed in other tickets. As part of this ticket, the AC below needs to be validated and implemented if it does not exist._ 

**If info Activate Upon Payment = Yes, then after the extension invoice is paid, the system should:**

1. Update folder status to Active
2. Update folder expiration to the value in Process Info 'New End Date' in ROW Admin process.

- The folder Issue Date and the folder Issue User should NOT be changed when the above functionality occurs. Stakeholders mentioned that this seems to be happening in some RW-TURP folders. This should be verified as part of this ticket. 

**If info Activate Upon Payment = No, then after the extension invoice is paid, the system should do nothing.**

**Add 2 New Infos (Extension start (75993) and Extension end (75994) dates)**
 - The New Infos (Extension start (75993) and Extension end (75994) dates) should updated by the process info
 - The two new info fields: 
    -  (Extension Start Date (75993) 
    -  (Extension End Date (75994)
----


### COS Payment Automation - EX Extension Fees #7284:

**Cannot begin working until #7631 (Update the EX extension workflow)**

**Acceptance Criteria**

- The AC provided below applies for EX folders with the following subtypes ONLY: 
-- Austin Energy
-- ATD
-- Austin Water
-- Corridor Program Office
-- Project Connect
-- Public Works
-- Watershed Protections
- When user makes attempt 'Extension Granted' in ROW Admin Process on EX folders, system should perform the following actions: 
-- Insert and bill fees based on fee logic (#7631).
-- Automatically Pay any Excavation Permit Fee - Secondary, Excavation Permit Fee - Extension and ATD Tech Surcharge fees that have been inserted and billed in the folder.
-- Payment type for this auto-payment should be recorded as Cost of Service.
-- The stamp user for the auto-payment should be AMANDA.
-- The paidby field for the auto-payment should be AMANDA.
----



### BUG: EX Date Extension Features need to NOT update the Extension Date when the Extension is Approved #7909:

If the EX work types are Franchise, ATD, Austin Energy, Austin Water, Corridor Program Office, Public Works, Watershed Protection when user Grants Extension:

Stakeholder: Kim and Paloma

Acceptance Criteria:
1. User saves "Extension Granted" Attempt
2. System should **NOT** update the Issue date with the Extension If the issuedate is already there
3. System updates **folder infos** extension start date and extension end date
4. System updates folder status to Active and folder expiration date to info extension end date
5. If there is no issue date & its a first time issue, it should write the issue date.

Note: Kim & Paloma just told her that they never want the issue date to change w/ the extension process on both EX & RW.

**Comment by TracyLinder** [comment]
### My updates are done
_FYI @Sreelatha-BlazeKannan @Nadin-Nader @vschandramouli_
- I updated the issue name & issue description to include 7909 & 7284. If you need to see previous AC it's in the comments above.

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - 
Need clarification on this, thought about this when I was testing #7631 changes.
For Subcodes Private Development/Project Connect In flow diagram, it says that do not create bill; but as per coding when activate upon payment info is Yes folder should become active and expiry date must be updated and folder description must be updated. But if the bill not created and not paid, can I make the folder status "Active"????

**Comment by TracyLinder** [comment]
@Sreelatha-BlazeKannan 

What will happen is that staff will manually create the bill. Then the customer pays the bill. Then that triggers activate upon payment. So then follow the activate upon payment flow. 


![AMANDA Misc. - Activate Upon Payment.jpg](https://images.zenhubusercontent.com/5cca1e522b8dfc0ee14dedf5/7a4bf90a-fadf-41e5-af1d-6741fd462d70)

**Comment by Sreelatha-BlazeKannan** [comment]
@TracyLinder - In Activate Upon payment flow - The yellow highlighted question, "How does the system know which expiration date to choose? " - I have updated the code to pickup the original date when extension date is NULL if not it will pickup extension date. Correct me if I am wrong.

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - As per the above understanding, Development is completed. Testing is in Progress,

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - Testing is in progress

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder  - Submitted DEV DP as was not able to test completely in localhost; having issues in a recreation of folders. Once it is moved to DEV, I will have to do one more complete round of testing and create a demo for review.  Please create a task id for me based on the commit id's below.

here are the commit id's :
1.	7c8fde7c5f2e 
2.	535dbe4cf59f

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan I have sent this to the DevOps CM Team to deploy to **_DEV_** under **SCTASK0190464**

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan this has been deployed to **_DEV_** Please UAT

**Comment by Nadin-Nader** [comment]
**Status:**
@Sreelatha-BlazeKannan is UATing this on DEV

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - UAT'd in DEV and found that valid lookup changes and auto payment was not working as expected; submitted the relevant changes and recommitted again. here is the commit id: efa6e3c68e2e; please approve and create task id for me.

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan I have sent this to the DevOps CM Team to deploy to **_DEV_** under **SCTASK0191917**

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan this has been deployed to DEV.

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - #7631 DEV DP submitted. I am submitting after Subha’s commit(#6255) due to dependency; Kindly approve and create task id for me.

Commit ID: 5b336005ac42

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan this has been sent to the DevOps CM Team to deploy to **_DEV_** under **SCTASK0194543**

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan  This has been deployed to **DEV** Please UAT

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - Submitted TEST DP’s for #7631 Card. Please approve and create a task id for me.

Here is the order of Commit IDs to promote; However #6255 Task ID: SCTASK0194755 must be promoted before these commits.

Order 1 Commit ID – 7c8fde7c5f2e & 535dbe4cf59f
Order 2 Commit ID - efa6e3c68e2e
Order 3 Commit ID -5b336005ac42

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan  This was sent to the devOps cm team to deploy to TEST under SCTASK0194793 and i snow deployed. Please UAT

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder  - Submitted PROD DP’s for #7631. Please approve and create task id for me to promote these commit ID’s to PROD.

Here is the order for PROD Promotion. However please make sure #6255 commit goes first to PROD as it is a dependency card for 7631.

Order 1 Commit ID’s - 7c8fde7c5f2e & 535dbe4cf59f 
Order 2 Commit ID - efa6e3c68e2e
Order 3 Commit ID - 5b336005ac42
Order 4 Commit ID - 51262e0d3a9f

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan this has been sent to the DevOps CM Team to be deployed to **_PROD_** under **SCTASK0194859**

**Comment by Nadin-Nader** [comment]
@Sreelatha-BlazeKannan this has been deployed to PROD

**Comment by TracyLinder** [comment]
**❌ UAT FAILED ❌**
In both my test cases multiple issues were found with the EX extension process. (My test notes are [here](https://github.com/cityofaustin/atd-data-tech/issues/8319). Confirming that this needs to be redeployed. 😫

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder - Tested in TEST and its working as expected. and ready for PROD Move.

**Comment by TracyLinder** [comment]
### UAT Passed. Please promote to PROD

**Comment by Sreelatha-BlazeKannan** [comment]
@Nadin-Nader , @TracyLinder  - Please deploy this to PROD only when #6255 moved.

**Comment by TracyLinder** [comment]
### PROD UAT

- I tested this today and it worked great. The only problem is with the folder expiration date does not populate, but we will fix that on another issue (not sure which one, maybe #8671. 

@Nadin-Nader can we close this?

**Comment by Nadin-Nader** [comment]
This has passed UAT. Closing
